src/map.cpp

```diff
@@ -4565,91 +4565,128 @@ void map::debug()
-       real_max_sight_range = distance_to_look;
-   }
+    update_visibility_variables();
 
-   if ((bio_night_active && dist < 15 && dist > natural_sight_range) || // if bio_night active, blackout 15 tile radius around player
-       dist > real_max_sight_range ||
-       (dist > light_sight_range &&
-         (lit == LL_DARK ||
-         (u_sight_impaired && lit != LL_BRIGHT) ||
-          !can_see))) {
-    if (u_is_boomered)
-     mvwputch(w, realy+getmaxy(w)/2 - center.y, realx+getmaxx(w)/2 - center.x, c_magenta, '#');
-    else
-         mvwputch(w, realy+getmaxy(w)/2 - center.y, realx+getmaxx(w)/2 - center.x, c_dkgray, '#');
-   } else if (dist > light_sight_range && u_sight_impaired && lit == LL_BRIGHT) {
-    if (u_is_boomered)
-     mvwputch(w, realy+getmaxy(w)/2 - center.y, realx+getmaxx(w)/2 - center.x, c_pink, '#');
-    else
-     mvwputch(w, realy+getmaxy(w)/2 - center.y, realx+getmaxx(w)/2 - center.x, c_ltgray, '#');
-   } else if (dist <= u_clairvoyance || can_see) {
-    drawsq(w, g->u, realx, realy, false, true, center.x, center.y,
-           (dist > low_sight_range && LL_LIT > lit) ||
-           (dist > sight_range && LL_LOW == lit),
-           LL_BRIGHT == lit);
-   } else {
-    mvwputch(w, realy+getmaxy(w)/2 - center.y, realx+getmaxx(w)/2 - center.x, c_black,' ');
-   }
-  }
- }
-    g->draw_critter( g->u, center );
+    lit_level ll;
+
+    for  (int realx = center.x - getmaxx(w)/2; realx <= center.x + getmaxx(w)/2; realx++) {
+        for (int realy = center.y - getmaxy(w)/2; realy <= center.y + getmaxy(w)/2; realy++) {
+            ll = apparent_light_at(realx, realy);
+            switch (ll) {
+                case LL_DARK: // can't see this square at all
+                    if (u_is_boomered)
+                        mvwputch(w, realy+getmaxy(w)/2 - center.y, realx+getmaxx(w)/2 - center.x, c_magenta, '#');
+                    else
+                        mvwputch(w, realy+getmaxy(w)/2 - center.y, realx+getmaxx(w)/2 - center.x, c_dkgray, '#');
+                    break;
+                case LL_BRIGHT_ONLY: // can only tell that this square is bright
+                    if (u_is_boomered)
+                        mvwputch(w, realy+getmaxy(w)/2 - center.y, realx+getmaxx(w)/2 - center.x, c_pink, '#');
+                    else
+                        mvwputch(w, realy+getmaxy(w)/2 - center.y, realx+getmaxx(w)/2 - center.x, c_ltgray, '#');
+                    break;
+                case LL_LOW: // low light, square visible in monochrome
+                case LL_LIT: // normal light
+                case LL_BRIGHT: // bright light
+                    drawsq(w, g->u, realx, realy, false, true, center.x, center.y, ll==LL_LOW, ll==LL_BRIGHT);
+                    break;
+                case LL_BLANK:
+                    mvwputch(w, realy+getmaxy(w)/2 - center.y, realx+getmaxx(w)/2 - center.x, c_black,' ');
+                    break;
+                default: // shouldn't happen
```

Comment:
> shouldn't happen

Than don't add it (-: No, really. The compiler would have warned you if you had forgotten one of the possible enum values (or if somebody adds another value and does not handle it here). However, with the default entry, this helpful feature of the compiler is completely disabled.

Having no code for the default will also create visible effect whenever this case actually happens (the displayed map will have "holes" where nothing is drawn), granted your magenta '!' is also very easy to spot.

Other than that, this commit looks fine.